src/utils/refreshDockerode.ts

```diff
@@ -0,0 +1,99 @@
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { DockerOptions } from 'dockerode';
+import Dockerode = require('dockerode');
+import { ext } from '../extensionVariables';
+import { addDockerSettingsToEnv } from './addDockerSettingsToEnv';
+import { exec } from './exec';
+import { isWindows } from './osUtils';
+
+const unix = 'unix://';
+const npipe = 'npipe://';
+
+const SSH_URL_REGEX = /ssh:\/\//i;
+
+// Not exhaustive--only the properties we're interested in
+interface IDockerEndpoint {
+    Host?: string;
+}
+
+// Also not exhaustive--only the properties we're interested in
+interface IDockerContext {
+    Endpoints: { [key: string]: IDockerEndpoint }
+}
+
+/**
+ * Dockerode parses and handles the well-known `DOCKER_*` environment variables, but it doesn't let us pass those values as-is to the constructor
+ * Thus we will temporarily update `process.env` and pass nothing to the constructor
+ */
+export async function refreshDockerode(): Promise<void> {
+    const oldEnv = process.env;
+    try {
+        process.env = { ...process.env }; // make a clone before we change anything
+        addDockerSettingsToEnv(process.env, oldEnv);
+        ext.dockerodeInitError = undefined;
+        ext.dockerode = new Dockerode(await getDockerodeOptions());
+    } catch (error) {
+        // This will be displayed in the tree
+        ext.dockerodeInitError = error;
+    } finally {
+        process.env = oldEnv;
+    }
+}
+
+async function getDockerodeOptions(): Promise<DockerOptions | undefined> {
+    // By this point any DOCKER_HOST from VSCode settings is already copied to process.env, so we can use it directly
+
+    try {
+        if (process.env.DOCKER_HOST &&
+            SSH_URL_REGEX.test(process.env.DOCKER_HOST) &&
+            !process.env.SSH_AUTH_SOCK) {
+            // If DOCKER_HOST is an SSH URL, we need to configure SSH_AUTH_SOCK for Dockerode
+            // Other than that, we use default settings, so return undefined
+            process.env.SSH_AUTH_SOCK = await getSshAuthSock();
+            return undefined;
+        } else if (!process.env.DOCKER_HOST) {
+            // If DOCKER_HOST is unset, try to get default Docker context--this helps support WSL
+            return await getDefaultDockerContext();
+        }
+    } catch { } // Best effort only
+
+    // Use default options
+    return undefined;
+}
+
+async function getSshAuthSock(): Promise<string | undefined> {
+    if (isWindows()) {
+        return '\\\\.\\pipe\\openssh-ssh-agent';
+    } else {
+        // On Mac and Linux, if it isn't set there's nothing we can do
+        // Running ssh-agent will yield a new agent that doesn't have the needed keys
+        throw new Error('In order to use an SSH DOCKER_HOST on OS X and Linux, you must configure an ssh-agent.');

```

Comment:
This error will be thrown and immediately eaten, we should instead show a warning/error message.